### PR TITLE
[v6r18][please merge quickly] Document Low level client classes

### PR DESCRIPTION
**Pull request by chaen**
"Always code as if the person who ends up maintaining your code is a violent psychopath who knows where you live." Luckily, I do not know the address ... :-)

This is just pure documentation of how the client side works when you perform an RPC call. probably more will come.
I also introduce a change in the config of sphinx to add the graphviz module. I've checked, and it is supported by readthedoc (http://docs.readthedocs.io/en/latest/builds.html#packages-installed-in-the-build-environment)

As a general note: we have a problem with the images we put in the doc. They are PNG, and once they are in, no one has anymore the original file to modify it. If we use a descriptive language like dot directly inside the documentation, anyone can modify it very easily. So I propose we use that when making class diagrams for example. 

Finally, it turned out to be extremely instructive to document that ! And I do have a few questions:
* What the hell is this ThreadConfig class ? It has the very strange concept to me of being a threading.local and yet a singleton !! That does not mean anything to me :-) Moreover, I cannot see it used anywhere. 
* What are the DIRAC Gateways ? Are they used at all ? I seem to have understood that it is a way to forward calls. 
* I am under the very strong impression that ```DIRAC.Core.Base.Client``` is not thread safe nor reentrant ! 
* There actually is a hell lot of options one can pass when constructing the client ! :-) So to perform a call with a server certificate or with another proxy does not require to change the environment ! That is a discovery to me ! :-) 

Can you please merge it quickly ? I would like to modify slightly the behavior of it, and I don't want to have merge conflicts. This PR contains only documentation

### CODE REVIEW

File: Core/DISET/private/BaseClient.py [hunk 1]
```diff
@@ -41,6 +42,23 @@ class BaseClient(object):
   __threadConfig = ThreadConfig()
 
   def __init__( self, serviceName, **kwargs ):
+    """
+      :param useCertificates: If set to True, use the server certificate
```
Comment (fstagni) [inline Core/DISET/private/BaseClient.py]: `:param serviceName` is missing.

**Comment by fstagni** [comment]
I guess the best person to approve this would be @andresailer 

Regarding your questions, we may try our luck with CCing @acasajus !

**Comment by zmathe** [comment]
I was the lucky person who had to work with the ThreadConfig when I introduced the concurrent.futures to web app. So, I can say a few word about ThreadConfig. In the first look and looking the code it seams it is not used. But it is not true. I would say this is a very cleaver and very simple solution for using RPC calls inside the web framework. What I mean is that we have to able to pass the information to the BaseClient inside the WebHandler:
```python
from DIRAC.Core.DISET.RPCClient import RPCClient
  .....
  @asyncGen
  def web_getJobData( self ):
    RPC = RPCClient( "WorkloadManagement/JobMonitoring", timeout = 600 )
    req = self._request()

    result = yield self.threadTask( RPC.getJobPageSummaryWeb, req, self.globalSort , self.pageNumber, self.numberOfJobs )

.....
```

So, this solution allow to pass the all information what is needed for make a call to a DIRAC service. The time when you made the RPC call from the web server, the web server host certificate is used with the added extra information using the ThreadConfig. In other word when we use the web server we have to able to provide the extra parameters which will be used for authorisation in the DIRAC service side.

**Comment by chaen** [comment]
@zmathe, I am sorry but I do not understand your example.
Where is the ```ThreadConfig``` here ? So you are saying that this class allows to pass extra information to the remote service ? So we use the web server certificate to talk to the DIRAC service, but we also tell to the dirac service who did the request ?

**Comment by chaen** [comment]
And btw, for crawling through it, managing to understand it, and not document it, you deserve to be GEOC for 2 months !!! :D

File: Core/DISET/RPCClient.py [hunk 2]
```diff
@@ -3,41 +3,100 @@
...
+  """ This class contains the mechanism to convert normal calls to RPC calls.
+      When instanciated, it creates a :any:`~DIRAC.Core.DISET.private.InnerRPCClient.InnerRPCClient` as an attribute.
+      Any attribute which is accessed is then either redirected to InnerRPCClient if it has it,
+      or creates a MagicMethod object otherwise. If the attribute is a function, MagicMethod will
+      trigger the RPC call, using the InnerRPCClient.
+
+      The typical workflow looks like this:
+
+
+      rpc = RPCClient('DataManagement/FileCatalog')
```
Comment (andresailer) [inline Core/DISET/RPCClient.py]: Should this be indented as a code block?

Comment (chaen) [inline Core/DISET/RPCClient.py]: Yes it is meant to be a code block like text. Can I indent it just like I do for rst ? Or how am I supposed to do ?

File: Core/DISET/private/BaseClient.py [hunk 3]
```diff
@@ -274,6 +374,13 @@ def __selectUrl( self, notselect, urls ):
 
 
   def __checkThreadID( self ):
+    """
+      CAUTION: just guessing....
```
Comment (andresailer) [inline Core/DISET/private/BaseClient.py]: There is a ..warn:: block you could use

File: Core/DISET/private/BaseClient.py [hunk 4]
```diff
@@ -304,57 +420,101 @@ def _connect( self ):
...
       gLogger.exception(lException = True, lExcInfo = True)
       return S_ERROR( "Can't connect to %s: %s" % ( self.serviceURL, repr( e ) ) )
+    # We add the connection to the transport pool
     trid = getGlobalTransportPool().add( transport )
+
     return S_OK( ( trid, transport ) )
 
   def _disconnect( self, trid ):
+    """ Disconnect the connection.
+        :param trid: Transport ID in the transportPool
```
Comment (andresailer) [inline Core/DISET/private/BaseClient.py]: Newline before `:param`

Comment (chaen) [inline Core/DISET/private/BaseClient.py]: Can't we have a post commit or post push hook to verify that kind of things ? Or even a test travis style ?

File: Core/DISET/RPCClient.py (hunk 2 above)
Comment (andresailer) [inline Core/DISET/RPCClient.py]: Yes, basically indent 2 spaces, add a second colon
```
      The typical workflow looks like this::


        rpc = RPCClient('DataManagement/FileCatalog')
  
        # Here, func is the ping function, which we call remotely.
        # We go through RPCClient.__getattr__ which returns us a MagicMethod object
        func = rpc.ping
  
        # Here we call the method __call__ of the MagicMethod
        func()
```

File: Core/DISET/private/BaseClient.py (hunk 4 above)
Comment (andresailer) [inline Core/DISET/private/BaseClient.py]: I am afraid sphinx doesn't produce a warning for this case.
And most private functions or classes do not get documenation created at the moment.

**Comment by chaen** [comment]
So after discussion with Zoltan and doing some debugging on this class, I understand more what it does.
As Zoltan says, this class allows to pass extra information when performing a call, in particular the delegated DN/Group/Setup/VO. To be used, the host performing the request has obviously to have the TrustedHost capabilities.  
The interesting bit is that this class is a Singleton, so we us the same object within one process, however the attributes of this class are thread locals ! Quite interesting... 
Now, I am not sure about why only this class has the threading.locals
After all, the RPCClient would probably deserve that too. 
I'll update the documentation

**Comment by chaen** [comment]
Okay I extended the doc

**Pull request merged**